### PR TITLE
Refactor layer indices and coords

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR refactors the indices and coordinates on each of the layers to reduce code duplication across layers (by putting common functionality on the base layer) and reducing the use of the `layer.viewer` property so that it will be easier to remove the `viewer` property completely from the individual layers in the future. The `viewer` property is now only used on the `BaseLayer`.

This PR accomplishes the following

- Removes all references to `viewer.dims.indices` from the individual layers. Instead there is now a `layer.indices` property on the `BaseLayer` that gets set in the viewer when the dims change. The `layer.ndim` property is used to ensure that the right number of dims are being sent to each layer

- Consolidates all the transforms relating to the scaling transformations from screen coordinates to image coordinates into a `layer.rescale` property, a `layer.cursor_position`, and a `layer.coordinates` all on the `BaseLayer`. This reduces duplicate code between the layers and was flagged in #193.

- Consolidates the `_refresh` and `update` methods onto the BaseLayer as they were almost identical across layers. The critical method that now differs across layers and gets called during that process is `set_view_slice`. `update`, `refresh`, `set_view_slice` might be able to be consolidated even further amongst themselves? Maybe @kne42 has ideas there.

The PR adds no new functionality. Although it is complete and ready for review now, this PR should probably wait until after #246 is merged. They are mostly independent, but there is one line that will have to change in this PR after that one goes in or visa-versa.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

- [x] run examples

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/components/_layers_list/model.py
Comment (sofroniewn) [inline napari/components/_layers_list/model.py]: This line will change after #246 goes in

File: napari/layers/_base_layer/model.py
Comment (kne42) [inline napari/layers/_base_layer/model.py]: what's the point of this? if it does nothing with the information that the value is `None`, shouldn't it just be an invalid value?

File: napari/layers/_base_layer/model.py
Comment (kne42) [inline napari/layers/_base_layer/model.py]: this name implies that this is a function that rescales the layer. would recommend that you choose a noun instead

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [inline napari/layers/_shapes_layer/model.py]:
```suggestion
                r = self._rotation_handle_length * self.rescale
```

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [inline napari/layers/_shapes_layer/model.py]:
```suggestion
            vertices = centers + self.rescale * self._highlight_width * offsets
```

File: napari/components/_layers_list/model.py
Comment (sofroniewn) [reply napari/components/_layers_list/model.py]: #246 has now been merged and this has been fixed

File: napari/components/_viewer/model.py
Comment (kne42) [inline napari/components/_viewer/model.py:243]: what's our plan when we implement new slicing operations?

**Comment by sofroniewn** [comment]
The checks for this PR are currently failing for the `labels-2d.py` example, but when I run `python examples/labels-2d.py` everything looks fine. If I run `pytest .` and test all the examples locally that specific test also fails. Any idea @kne42  what could cause a difference between the `pytest .` and the `python examples/labels-2d.py`?

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: can you elaborate? right now the value passed will sometimes be none because vispy does that (?!) so I just wanted it to be ignored and not update the `cursor_position` value

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: how about `scale_factor`?

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [reply napari/layers/_shapes_layer/model.py]: fixed

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [reply napari/layers/_shapes_layer/model.py]: fixed

File: napari/components/_viewer/model.py
Comment (sofroniewn) [reply napari/components/_viewer/model.py:243]: i think we'll need to send more info over to the layers, but I like setting properties like this - or do you think this is an anti-pattern?

**Comment by kne42** [comment]
it fails on `shapes_to_labels`, not `labels-2d`, and it does so for both python and pytest on my machine

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: I will move this check outside the setter

File: napari/components/_viewer/model.py
Comment (kne42) [reply napari/components/_viewer/model.py:243]: not sure @jni do u wanna weigh in?

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: fixed

File: napari/components/_dims/model.py
Comment (jni) [inline napari/components/_dims/model.py:151]: :blush: :blush: :blush:

File: napari/components/_viewer/model.py
Comment (jni) [reply napari/components/_viewer/model.py:243]: I think the properties can be overwritten by the layers, and then they can do what they want, so it *should* be fine? ie in terms of expressivity they are equivalent. I'm certainly happy for us to leave this as is for now and revert *if* it becomes a problem.

File: napari/layers/_base_layer/model.py
Comment (jni) [inline napari/layers/_base_layer/model.py]: I suggest:

```
2-tuple of cursor coordinates in image (pixel) space.

The setter expects the coordinates in canvas (real) space and transforms them
to image space, which is how they are stored internally.
```

or vice versa, I'm not sure what's happening, but that sort of stuff. =)

File: napari/layers/_base_layer/model.py
Comment (jni) [inline napari/layers/_base_layer/model.py]: Can you add a note about the 2D-only nature of this? I'm puzzled about why they are leading rather than lagging dimensions here.

File: napari/layers/_base_layer/model.py
Comment (jni) [inline napari/layers/_base_layer/model.py]: In this docstring, or maybe as comments, it is again worthwhile to explain the different coordinate conventions between vispy and NumPy, and that that explains the swapping of the dimension ordering in the following lines.

Comment (jni) [reply napari/layers/_base_layer/model.py]: And, it might be a good idea, whenever coordinates are mentioned *anywhere* to specify whether they are in canvas units and/or ordering, or in pixel/array units and/or ordering.

Comment (jni) [reply napari/layers/_base_layer/model.py]: ... And perhaps we should aim to have a *single* place where the two are converted, as currently it seems to happen all over the place, which might make refactors and modifications (such as arbitrary ordering/display) harder than necessary. Does this seem feasible to you in the short term?

File: napari/layers/_base_layer/model.py
Comment (jni) [inline napari/layers/_base_layer/model.py]: I feel like this should be either the same in each dimension, in which case, rather than the mean, just pick one, or allowed to be different in each dimension, in which case, we should return both.

Comment (jni) [reply napari/layers/_base_layer/model.py]: Also, the fact that it depends on the current zoom level suggests one more set of coordinates: screen coordinates. Or does zoom mean level of detail here for multiscale pyramids?

File: napari/layers/_image_layer/model.py
Comment (jni) [inline napari/layers/_image_layer/model.py:179]: I'm confused about why the `ndim` slicing is no longer needed, and ditto for the indices argument. It might be more prudent to give indices a default value of `None` and then write:

```python
indices = indices or self.indices
```

File: napari/layers/_image_layer/model.py
Comment (jni) [inline napari/layers/_image_layer/model.py:334]:
```
coord = np.round(self.coordinates).astype(int)
coord[-2:] = np.clip(coord[-2:], 0, np.asarray(self.image_view.shape) - 1)
```

As mentioned elsewhere, it'd be good here to mention the coordinate space when you use it.

File: napari/layers/_image_layer/model.py
Comment (jni) [inline napari/layers/_image_layer/model.py]: Again, please note the coordinate space used for each variable.

File: napari/layers/_labels_layer/model.py
Comment (jni) [inline napari/layers/_labels_layer/model.py]: The changes to this file make me wonder whether Labels should be a subclass of Image, rather than a sister implementation. Thoughts? (This can obviously wait for a later PR, it's just a good time to have the discussion I think.)

File: napari/layers/_labels_layer/model.py
Comment (jni) [inline napari/layers/_labels_layer/model.py]: As above.

File: napari/layers/_labels_layer/model.py
Comment (jni) [inline napari/layers/_labels_layer/model.py]: Why is more rounding and casting needed here?

**Review by jni — DISMISSED** [review]
@sofroniewn good job on keeping all this stuff straight! =P I've made some comments, mainly relating to helping the rest of us keep all the coordinate spaces straight in our heads also. I've identified at least three spaces:

- image (pixel) space
- canvas (real) space
- screen space

It would be good if each of those was annotated where used. Perhaps even a short doc page would be a very valuable thing at this point!

File: napari/components/_viewer/model.py
Comment (sofroniewn) [reply napari/components/_viewer/model.py:243]: ok great - sounds good

**Comment by sofroniewn** [comment]
@jni - i'm about to go through your comments, but just want to be crystal clear on the coordinate spaces as I really have 4 different ones that I can think of.

 If I have a 3-D dataset then I have the following 4 coordinates

`(x, y)` -> `(x', y')` -> `(z, y', x')` -> `(z, y'', x'')`

where (x, y) are integer valued in the space of my screen, (x', y') are floats in the 2D space of the raw data that is visible, (z, y', x') is now a tuple that refers to an actual point in the raw data volume and `x'' = int(round(x'))`, `y'' = int(round(y'))` are integer valued also in the space of the raw data, now refering to an actual pixel.

In this framing `(x, y) = event.pos` is not saved on the layer, but is the input to the setter of the `cursor_position`.  `(x', y') = layer.cursor_position`, `(z, y', x') = layer.coordinates`, and `(z, y'', x'')` doesn't have a name.

The setter of this is used to do the `layer.cursor_position` is used to do the transformation from `(x, y)` -> `(x', y')`, but `layer.cursor_position` is only used inside `layer.coordinates` and never used on its own. It strikes me that `layer.cursor_position` could easily be dropped in favor of just using the setter of `coordinates` itself. Is it strange that `coordinates` is not integer valued? Do we then want another property that is the corresponding pixel values `integer_coordinates = np.round(coordinates).astype(int)`, but if so what should it be called.

How do these concepts map onto screen space, canvas (real) space, image (pixel) space that you mentioned above?

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: i'm eliminating this property as it was not very useful

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: as mentioned in the other comment, i'm eliminating this property as it was more confusing than useful

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: yup, agreed - I'm cleaning this up, and will improve the doc strings

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: This is really screen coordinates, shouldn't have called it canvas. Zoom does not relate to the image pyramids directly (though you could use this value to determine which level of the pyramid to show). I've improved the doc string.

Also both values should be the same so I've just taken the first

File: napari/layers/_image_layer/model.py
Comment (sofroniewn) [reply napari/layers/_image_layer/model.py:179]: the ndim slicing was moved directly onto the base layer. I'm not really sure I see a need for passing custom indices here, though we could always add that in

File: napari/layers/_image_layer/model.py
Comment (sofroniewn) [reply napari/layers/_image_layer/model.py:334]: i've improved the doc string

Comment (sofroniewn) [reply napari/layers/_image_layer/model.py:334]: and made the change to clip

File: napari/layers/_image_layer/model.py
Comment (sofroniewn) [reply napari/layers/_image_layer/model.py]: `cursor_position` has now been eliminated, i'll add something everywhere about `event.pos`

File: napari/layers/_labels_layer/model.py
Comment (sofroniewn) [reply napari/layers/_labels_layer/model.py]: fixed as above

File: napari/layers/_labels_layer/model.py
Comment (sofroniewn) [reply napari/layers/_labels_layer/model.py]: fixed

File: napari/layers/_labels_layer/model.py
Comment (sofroniewn) [reply napari/layers/_labels_layer/model.py]: yes - it might make a lot of sense to have Labels as a subclass of Image. I think they have a lot in common and that you'd just need to overwrite a few properties / methods, customize a few others to make it work. I'm also thinking about how we'd extend image pyramid support to labels too, and making labels be a subclass and just adding image pyramid support to the main image layer would be one way.

**Comment by sofroniewn** [comment]
@jni I have removed the `cursor_position` property, now there is just the `coordinates` property. I also now only talk about two different coordinate systems - there is screen coordinates which are (x, y) tuples representing points on my screen (I guess now that I think about it, maybe we should call these canvas, as (0,0) is not the top left of my entire screen, but it is the top left of the vispy canvas, which is the same as the top left of the QtWidget that corresponds to the canvas - I can change this to canvas if you want) - and there are image coordinates, which represent points (not necessarily pixel valued) in the coordinates of the raw data. These can be n-D and are ordered according to our napari ordering conventions.

I've tried to make this more clear / improve the doc strings - let me know if you have further requests for changes

File: napari/layers/_labels_layer/model.py
Comment (jni) [reply napari/layers/_labels_layer/model.py]: Ok, we'll leave that to another PR, but glad we are in agreement. =)

Comment (sofroniewn) [reply napari/layers/_labels_layer/model.py]: we could even think about making a common class for all our mesh layers (i.e. shapes, markers, vectors) but i don't think we're ready for that yet

**Comment by jni** [comment]
@sofroniewn I don't think we have the same definitions of screen coordinates. =) canvas is right if (0, 0) refers to the top left corner of the vispy canvas. If (0, 1) does not correspond to "1 monitor pixel right of (0, 0) on your screen", then it is not screen coordinates, it is canvas coordinates.

From what I can tell, we are not dealing with screen coordinates at all.

One possible point of confusion is that currently our canvas coordinate spacing is the same as the pixels in the image. But as I recall we had a chat about having the canvas coordinates represent "real" space, in the sense that if the image has metadata specifying a scale, then we draw it taking that scale into account, and canvas coordinates would now be e.g. microns. In the status bar, we probably want to include both pixel and real space coordinates, or have a very easy toggle to switch between them (it is annoying in Fiji that this is not always easy to do). 

I'm not super-clear on how transposition fits into this scheme. That can be very confusing, because you don't want to transpose things arbitrarily. I recall a horror story about doctors getting the side of a surgery wrong because of a bug in the volumetric image display software.

**Comment by sofroniewn** [comment]
Yikes! Yes - we're not dealing with screen coordinates at all. I have now replaced all mentions of screen coordinates with canvas coordinates. In my terminology though the size of a pixel in canvas coordinates is the same as the size of a pixel in screen coordinates though and is fixed, regardless of the zoom on the vispy camera. The transformation to image coordinates then is really about the conversion through the zoom and position of the vispy camera. I havn't made any consideration yet of "real" coordinates like microns as we havn't done anything with any metadata there yet, but I'd suggest that we locate that conversion to a single place, that lives elsewhere on the base layer and is not part of this PR

**Comment by jni** [comment]
Hmm, ok, the terminology needs refinement then. Let's add another coordinate space!

Screen > Viewport > Canvas > Array

Viewport is the thing you describe and it is related to screen coordinates only in translation of the origin. (And/or inversion of an axis and/or transposition, not sure?)

Now, from what you're telling me, currently you're assuming Array and Canvas are the same thing and that conversion isn't happening anywhere, right?

**Comment by sofroniewn** [comment]
Hmm @jni, looking through the [vispy scene docs](http://vispy.org/scene.html) they use a term `viewbox` but that refers to something not in screen / viewport coordinate as how you use them above. I'm not sure about the term `viewport`, we just have so many viewer / view etc being used.

I will note though that in the vispy docs they refer to the `event.pos` on the mouse event property as follows

```python
pos : (int, int)
    The position of the mouse (in screen coordinates).
```

This is the input to the `layer.coordinates` setter - but I will note that the `(0, 0)` is not in the top left of the screen, but the top left of what you call the `viewport` see the screenshot
image
We could say in the coordinates of the "Canvas widget" as we do refer the that thing as the `canvas` on the `QtViewer`.

You seem to use the word canvas to be after the transformation back to what I call "image", I also find that confusing, but that is the space that the `layer.coordinates` are in right now. For the "array" there is also a rounding and clipping that occurs in the image / labels layers, and I guess I havn't been giving that thing a special name yet, though it is the `coord` output of the `get_value` function

We might need to talk about this on thursday to avoid going around in circles!!

**Comment by jni** [comment]
> We might need to talk about this on thursday to avoid going around in circles!!

This is a great idea! :joy: A diagram would probably also be super useful — I'll see if I can produce one in time.

**Comment by jni** [comment]
For tomorrow:

My Sketches - 2019-05-09 19 49 51

The main source of confusion, I think, is that there's three coordinate systems that share the same origin. (?):

- the `event.pos` coordinates, which are pixels on the monitor. Depending on the zoom and pan state, `event.pos` will point to different bits of the data. (Or is this origin at the top left corner of the black rectangle? I can't tell where the mouse cursor is in your screenshot.)
- the data pixels. Are these axes transposed with respect to `event.pos`?
- a "real" coordinate space. We haven't yet used this, but we will have to because of e.g. wanting to show registered layers one on top of the other. This means that images will not be just images but images + transforms, and so we cannot always count on the image pixels being aligned with the screen coordinate space.

The transformation between real coordinates and data pixel coordinates remains the same regardless of what we do the view (zoom/pan/rotate, whatever). In *most* cases, it will be a simple scaling.

**Comment by sofroniewn** [comment]
@jni here's a screenshot defining things how I see it, and making it clear what's in `event.pos`

image

**Comment by sofroniewn** [comment]
@jni if you can suggest changes to the doc strings based on the figure I posted above I will make them and then this should be ready to merge. Thanks!

**Comment by sofroniewn** [comment]
So the PR as it is now refers to "canvas coordinates" that are (x, y) tuples relative to the top left pixel of the canvas on the screen in units of the screen. After the transform the PR refers to "image coordinates" which are in the space of the image and have the length as the same number of dimensions as the layer and in the ordering of the layer. They aren't necessarily integer valued or clipped.

I think this is clear for now, and I'm going to ask @kne42 to give it one more look over so we can merge and move on with the refactor. If @jni wants the naming further clarified we can make follow up PRs modifying the doc strings

File: napari/layers/_base_layer/model.py
Comment (kne42) [inline napari/layers/_base_layer/model.py]: so like is this a `tuple of slice`? what's the actual type? what does it mean? (see below for an example of a more clear docstring)

File: napari/layers/_base_layer/model.py
Comment (kne42) [inline napari/layers/_base_layer/model.py]: do not include data type in description. it should be separate (and more clear)

```suggestion
        """tuple of float: Coordinates of the cursor in the respective image space of each layer.
```

File: napari/layers/_image_layer/model.py
Comment (kne42) [inline napari/layers/_image_layer/model.py]:
```suggestion
        value : int, float, or sequence of int or float
```

File: napari/layers/_image_layer/model.py
Comment (kne42) [inline napari/layers/_image_layer/model.py]: Docstring summaries/parameter descriptions should not redundantly list what type it returns/takes if possible.

```suggestion
        """Generate a status message based on the coordinates and information about
```

File: napari/layers/_labels_layer/model.py
Comment (kne42) [inline napari/layers/_labels_layer/model.py]: PEP8
```suggestion
        self.cursor_size = self._brush_size / self.scale_factor
```

File: napari/layers/_labels_layer/model.py
Comment (kne42) [inline napari/layers/_labels_layer/model.py]:
```suggestion
            self.cursor_size = self.brush_size / self.scale_factor
```
also may want to abstract this at some point

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [inline napari/layers/_shapes_layer/model.py]:
```suggestion
            r = self._rotation_handle_length * self.scale_factor
```

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [inline napari/layers/_shapes_layer/model.py]:
```suggestion
            r = self._rotation_handle_length * self.scale_factor
```

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [inline napari/layers/_shapes_layer/model.py]:
```suggestion
                sizes = self._vertex_size * self.scale_factor / 2
```

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [inline napari/layers/_shapes_layer/model.py]:
```suggestion
                sizes = self._vertex_size * self.scale_factor / 2
```

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [inline napari/layers/_shapes_layer/model.py]:
```suggestion
                    threshold = self._vertex_size * self.scale_factor / 8
```

File: napari/layers/_shapes_layer/model.py
Comment (kne42) [inline napari/layers/_shapes_layer/model.py]: im just gonna stop here. maybe we'll fix at a later point. all we really need is someone to run `autopep8`

you should hook up pep8/flake8 on your IDE to resolve these issues prior to push

**Review by kne42 — APPROVED** [review]
approved pending changes

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: Tried fixing, the type is either int or a Slice object - https://docs.python.org/3/library/functions.html#slice

File: napari/layers/_base_layer/model.py
Comment (sofroniewn) [reply napari/layers/_base_layer/model.py]: fixed

File: napari/layers/_image_layer/model.py
Comment (sofroniewn) [reply napari/layers/_image_layer/model.py]: fixed

File: napari/layers/_image_layer/model.py
Comment (sofroniewn) [reply napari/layers/_image_layer/model.py]: fixed

File: napari/layers/_labels_layer/model.py
Comment (sofroniewn) [reply napari/layers/_labels_layer/model.py]: fixed

File: napari/layers/_labels_layer/model.py
Comment (sofroniewn) [reply napari/layers/_labels_layer/model.py]: fixed

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [reply napari/layers/_shapes_layer/model.py]: fixed

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [reply napari/layers/_shapes_layer/model.py]: fixed

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [reply napari/layers/_shapes_layer/model.py]: fixed

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [reply napari/layers/_shapes_layer/model.py]: fixed

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [reply napari/layers/_shapes_layer/model.py]: fixed

File: napari/layers/_shapes_layer/model.py
Comment (sofroniewn) [reply napari/layers/_shapes_layer/model.py]: ok, i tried to fix all the rest of these - it is strange i didn't get any pep8 errors on these things

**Comment by sofroniewn** [comment]
I have fixed the remaining requests from @kne42 and will merge once the tests pass

**Pull request merged**
